### ISSUE TITLE
bemenu-run shows non-executable files in $PATH.

### ISSUE THREAD

**Issue by crocket** [Cloudef/bemenu#132]
Linux distribution packages install non-executable files in $PATH.

**Comment by Cloudef** [comment]
`$PATH` ideally should only contain executables, that's the purpose of it. Doing additional check whether file is executable or not could potentially make `bemenu-run` much slower as it would need `stat` system call for every file!

**Comment by crocket** [comment]
dmenu-run creates a cache through dmenu_path.

**Comment by Cloudef** [comment]
Yeah, you can optionally do that yourself as well. The `bemenu-run` is designed to read PATH as is and rely on linux's filesystem cache. So it's usually fast without any own caching.

**Comment by crocket** [comment]
For now, I pipe the output of dmenu_path to bemenu.

**Comment by Cloudef** [comment]
I could perhaps also introduce option for bemenu-run that filters the PATH.

**Comment by crocket** [comment]
What about caching?

**Comment by Cloudef** [comment]
If you want external cache, then using something like dmenu_path or rolling your own is recommended as it's also simpler than implementing it in bemenu-run, doing that you could even supply bemenu the shell's own cache. bemenu-run is simple program that only relies on the filesystem cache, doesn't write anything to the disk nor doesn't have to worry when to invalidate the cache.

**Comment by crocket** [comment]
Is bemenu-run going to be reasonably fast if it checks execution permission?

**Comment by Cloudef** [comment]
That I don't know without testing. I'd assume filesystem cache also caches metadata so my guess would be yeah.

**Comment by Earnestly** [comment]
It will be fast on a hot cache but can be extremely slow when it's cold depending on the filesystem and medium
